docs/non-normative/groups/system/design-philosophy.md

```diff
@@ -0,0 +1,259 @@
+- It is Linux exclusive, and is unlikely to ever be introduced in other
+  operating systems since the other major platforms have their own versions of
+  it (Windows Job Objects, BSD Jails, etc)
+
+This means we should pick a name that matches the verbiage used by specialists
+in context when referring to this concept. The way you would refer to this would
+be "a process's cgroup, collected from `/proc/<pid>/cgroup`". So we would start
+with the name `process.cgroup`. We also determined that this attribute is
+Linux-exclusive and are confident it will remain as such, so we land on the name
+`process.linux.cgroup`.
+
+Since this metric falls under Specialist Class, we don't want to be too
+prescriptive about the value. A user who needs to know the `cgroup` of a process
+likely already has a pretty good idea of how to interpret it and use it further,
+and it would not be worth it for this Working Group to try and come up with
+every possible edge case for how it might be used. It is much simpler for this
+attribute, insofar as it falls under our purview, to simply reflect the value
+from the OS, i.e. the direct value from `cat /proc/<pid>/cgroup`. With cgroups
+in particular, there is high likelihood that more specialized semconv
+instrumentation could be developed, particularly in support of more specialized
+container runtime or systemd instrumentation. It's more useful for a working
+group developing special instrumentation that leverages cgroups to be more
+prescriptive about how the cgroup information should be interpreted and broken
+down with more specificity.
+
+## Operating System in names
+
+Relevant discussions:
+[\#1255](https://github.com/open-telemetry/semantic-conventions/issues/1255),
+[\#1364](https://github.com/open-telemetry/semantic-conventions/pull/1364#discussion_r1852465994)
+
+Monitoring operating systems is an old practice, and there are lots of ways to
+skin the cat within different platforms. There are lots of metrics, even in
+basic stuff like memory usage, where there are platform specific pieces of
+information that are valuable to those who really specialize in that platform.
+
+Thus we have decided that any instrumentation that is:
+
+1. Specific to a particular operating system
+2. Not meant to be part of what we consider our most important general use cases
+
+will have the Operating System name as part of the namespace.
+
+For example, there may be `process.linux`, `process.windows`, or `process.posix`
+names for metrics and attributes. We will not have root `linux.*`, `windows.*`,
+or `posix.*` namespaces. This is because of the principle we’re trying to uphold
+from the [Namespaces section](#namespaces); we still want the instrumentation
+source to be represented by the root namespace of the attribute/metric. If we
+had OS root namespaces, different sources like `system`, `process`, etc. could
+get very tangled within each OS namespace, defeating the intended design
+philosophy.
```

Comment:
I'm curious what would be specific problems if we gave up on the prefix and use OS name as a root?

I'm trying to document naming patterns we have in https://github.com/open-telemetry/semantic-conventions/pull/1708

and I'm actually struggling to understand what benefit the `domain` prefix brings.

Reply:
E.g. what should I do if I want to describe a property of OS that's not about process or system? which namespace would I use?

Reply:
a few points:

1. The thing being reported is more important than instrumentation source. I hope that current instrumentations are temporary and we'll see more and more native ones. When you think about  native instrumentations, things change. E.g. 
    - windows itself emits tons of metrics/events. Should they all be reported under `os` ? Some of them describe system performance, some describe user behavior. Do we group all of them under `os`? It seems redundant. Or just some of them? How do we decides?
    - any database has management/control plane operations that has nothing to do with DB features (auth, permissions, connection management, scaling, etc). Do we report them as `db.{mydb}.*` attributes/metrics. E.g. [`db.cassandra.paxos.prepare.duration`](https://cassandra.apache.org/doc/4.1/cassandra/operating/metrics.html#:~:text=row%20cache%20misses.-,CasPrepare,-Latency) - protocol is orthogonal to the DB features of cassandra, but the metric is still about cassandra.  Now, do we report `db.cassandra.compaction.something` (because it's about database) and `cassandra.paxos.prepare.duration` because it's about protocol?

    TL;DR: any specific system/client lib has features that belong to more than one root namespace. How instrumentation is done may change (from specific collector component to native one), but metrics we define should survive it.

2. Having common root namespace for the "General Class" makes perfect sense to me: everything common about OS goes under `system`, everything common about databases goes under `db`. 

3. I'm challenging the "Specialist Class" naming: I'm reporting different metrics related to `jvm`, everyone who cares knows that `jvm` is a runtime, `runtime` in front of it is redundant. If I care about cassandra-specific metrics, I'm no longer in DB domain - `cassandra` is the root namespace


I.e. How strong do we feel about 

https://github.com/open-telemetry/semantic-conventions/blob/01f43e943767259ebb72f637b8fe5b428c8a16b0/docs/non-normative/groups/system/design-philosophy.md?plain=1#L107-L108

>  For example, a metric for a process's cgroup would be `process.linux.cgroup`,
   given that cgroups are a specific Linux kernel feature.

Vs `linux.cgroup` ?

Reply:
[UPDATE]:

The `process.linux.cgroup` seems to be tricky since `process` is meaningful there. cgroup is a property of the (linux) process.  I believe it's less tricky in most of other semconv cases. Let me see.

Reply:
I didn't cover well enough the point you brought up, which is that there may be a case where the platform/domain name becomes so specific and disparate in itself that it warrants being considered a unique instrumentation source/root namespace/category.

Taking the `db.cassandra` example, I imagine it would follow a process like this hypothetical scenario:
1. `db.*` may have a large number of attributes that can be used cross platform, and our guidance largely expects that they are used as much as possible. For stuff specific to Cassandra some `db.cassandra.*` stuff starts to get sprinkled into the namespace when necessary.
2. Overtime more Cassandra specialists join in and start adding more and more `db.cassandra` attributes. It begins to pollute the `db` category, such that the instrumenting Cassandra in particular has a far greater variety of things very specific to it. The fact that so many `db.cassandra.*` things are being added implies that these things are Cassandra specific already, since they should have been using generic attributes when possible.
3. A decision is made that since Cassandra-specific instrumentation has become so rich and disparate from everything else in the `db.*` namespace, it makes the most sense to split `cassandra` out into its own root namespace to represent the richness of the instrumentation available for it and how much it diverts.

Applying this thought process to the questions in your above comment:

> windows itself emits tons of metrics/events. Should they all be reported under os ?

I think the steps from above could still occur in the same order. We'd try to use as many generic attributes as possible, introducing `os.windows.*` where it is specifically necessary. If the `os.windows` category becomes so deep and disparate from everything else, then it might make sense to move it into its own namespace.

> How strong do we feel about `process.linux.cgroup` vs `linux.cgroup`

It is essentially for the reason you stated in your update, but this is a bit of a different case; whether Linux exclusive or not, the `process` will always be the instrumentation source. And where something like `os` and `db` may be more like categories than an explicit object being instrumented, `process` will always be the most important root namespace and `linux.cgroup` on its own wouldn't carry the same semantic meaning.

That being said, there is a case to be made for `cgroup` here, as `cgroup` is itself a potentially instrumentable source. It's (hopefully) only a matter of time before semantic conventions come along to instrument `cgroup`s in more detail. In that case, I wonder if that would start in `linux.cgroup` or if it would just be `cgroup` off the bat. That I'm not sure about.

---

The problem I foresee with my own ideas here is it assumes things can be moved around easier. I think the problem with my thought process is that it sort of requires foresight to make sure that there isn't the potential for us to want to extract a platform namespace before reaching the point of stability for another namespace. So all of this probably needs some more thought and discussion. Maybe when I come back from the holidays I'll have the answer (probably not but I can dream :smiley:).

Reply:
Given that some of the things we're discussing (`db.cosmosdb`) is reasonably close to stability, I think we should try to envision as many things as we can - we might not have another chance in the next few years 🤞 

I don't think we need a rigid naming policy - i.e. `process.linux.cgroup` can be whatever makes the most sense for it.
I'll bring up the general naming to Semonv SIG after the break.

Happy holidays!